File: Py/day-01.py
# --- STEP 1: Reading and Flattening the Data ---
# Open the file 'input.txt', read its entire content as a single string.
#
# The .split() method then splits the string by ALL whitespace (spaces and newlines),
# resulting in a list of string numbers.
#
# map(int, ...) converts every string in that list to an integer.
#
# The [*...] syntax (list comprehension with the unpacking operator)
# converts the resulting map object into a single, flat list of integers.
data = [*map(int, open("../inputs/input-01.txt").read().split())]

# --- STEP 2: Splitting, Sorting, and Assigning ---

# data[0::2] is a list slice:
#   - Starts at index 0 (the first element).
#   - Steps by 2 (takes every other element: 0, 2, 4, ...).
# This effectively collects all the numbers from the **first column** of the input.
#
# data[1::2] is also a list slice:
#   - Starts at index 1 (the second element).
#   - Steps by 2 (takes every other element: 1, 3, 5, ...).
# This effectively collects all the numbers from the **second column** of the input.
#
# sorted(...) sorts each resulting list numerically.
# The tuple assignment (A, B = ...) assigns the sorted lists to variables A and B.
A, B = sorted(data[0::2]), sorted(data[1::2])

# --- STEP 3: Printing the Results (Two Calculations) ---

# The print function will output the results of two separate calculations,
# separated by a comma (the default print separator).

print(
    # --- Calculation 1: Sum of Absolute Differences (total distance between lists) ---
    #
    # map(lambda a, b: abs(a - b), A, B):
    #   - The lambda function takes one element from A (as 'a') and one from B (as 'b')
    #     at the same index (since 'map' iterates over both lists in parallel).
    #   - It calculates the absolute difference: |a - b|.
    #   - Example: |31594 - 24099|, |46608 - 70524|, |78052 - 93577|
    # sum(...) adds up all these absolute differences.
    sum(map(lambda a, b: abs(a - b), A, B)),

    # --- Calculation 2: Sum of A-values Weighted by their Count in B (similarity score) ---
    #
    # map(lambda a: a * B.count(a), A):
    #   - The lambda function takes one element from A (as 'a').
    #   - B.count(a) counts how many times 'a' (the value from A) appears in list B.
    #   - It calculates: a * (count of 'a' in B).
    # sum(...) adds up all these weighted values.
    sum(map(lambda a: a * B.count(a), A)),
)

File: JS/day-01.js
/**
 * Processes the raw file input string into two separate arrays of numbers.
 * This function performs the necessary string manipulation and type conversion.
 * * @param {string} input - The raw string content read from the file.
 * @returns {Array<Array<number>>} An array containing two lists: [leftCol, rightCol].
 */
const formatLists = (input) => {
  // .trim(): Removes leading/trailing whitespace from the entire file content
  // .split("\n"): Splits the cleaned string into an array of individual lines (rows).
  const rows = input.trim().split("\n");

  // Initialize two arrays (lists) to hold the numbers from the respective columns.
  const leftCol = [];
  const rightCol = [];

  // Iterate over each line (row) read from the input file.
  for (let i = 0; i < rows.length; i++) {
    // .split(/\s+/): Splits the current row string by one or more whitespace characters.
    // This correctly handles multiple spaces or tabs between the two numbers.
    const [left, right] = rows[i].split(/\s+/);

    // .push(+value): Adds the value to the list. The unary plus operator (+)
    // is a fast, explicit way to convert a string ('left' or 'right') into a number.
    leftCol.push(+left);
    rightCol.push(+right);
  }
  // Return the two populated lists.
  return [leftCol, rightCol];
};

/**
 * Calculates the 'Total Distance' between the two columns. 
 * This requires sorting both lists and pairing them by index.
 * * @param {string} ids - The raw file input string.
 * @returns {number} The sum of absolute differences.
 */
const totalDistance = (ids) => {
  // Deconstruct the array returned by formatLists into the two column arrays.
  const [leftCol, rightCol] = formatLists(ids);

  // .sort((a, b) => a - b): Sorts the array numerically in ascending order (in-place).
  const leftSorted = leftCol.sort((a, b) => a - b);
  const rightSorted = rightCol.sort((a, b) => a - b);

  let distance = 0;
  // Iterate over the sorted lists. The length should be the same.
  leftSorted.forEach((id, i) => {
    // Get the corresponding value from the second list.
    const partner = rightSorted[i];
    
    // The absolute difference is calculated as (largest - smallest) to avoid using Math.abs
    const smallest = Math.min(id, partner);
    
    // Add the absolute difference to the running total.
    distance += largest - smallest;
  });
  return distance;
};

/**
 * Calculates the 'Similarity Score' or 'Weighted Sum' (Part 2). 
 * The logic is: Sum(element_A * count_of_element_A_in_B)
 * * @param {string} ids - The raw file input string.
 * @returns {number} The calculated similarity score.
 */
const idPow = (ids) => {
  // Read and format the data again. (Performance could be improved by passing the lists
  // directly from totalDistance, but this ensures function isolation).
  const [leftCol, rightCol] = formatLists(ids);
  
  // Initialize an object (hash map) to store the frequency of numbers in the second column.
  const numberCounts = {};
  
  // FIRST PASS: Populate the frequency map using the second column (rightCol).
  rightCol.forEach((num) => {
    // Check if the number already exists as a key in the map.
    if (num in numberCounts) {
      numberCounts[num]++;
    } else {
      numberCounts[num] = 1; // Initialize count to 1.
    }
  });

  let popularity = 0;
  // SECOND PASS: Calculate the weighted sum using the first column (leftCol).
  leftCol.forEach((num) => {
    // Get the count ('pow') of the current number 'num' from the map.
    // The || 0 ensures that if 'num' is not found in the map, 'pow' defaults to 0.
    const pow = numberCounts[num] || 0;
    
    // Add the element multiplied by its count in the other list.
    popularity += num * pow;
  });
  return popularity;
};

// --- MAIN EXECUTION BLOCK ---

const fs = require("fs");
// fs.readFileSync: Reads the file 'input.txt' synchronously into a string.
// "utf-8": Specifies the text encoding.
const input = fs.readFileSync("../inputs/input-01.txt", "utf-8");

// Calls both functions with the raw input string and logs the results to the console.
console.log(totalDistance(input), " and ", idPow(input));
